### ISSUE TITLE
Remove panic opportunities

### ISSUE THREAD

**Issue by joshlf** [google/zerocopy#1661]
These are the result of auditing (as of 000318497d86599e24cecb66b945c86a53af3163) for panic opportunities. Some have been left off which are either unavoidable, in progress of being removed (#1658), or downstream of ones listed here (namely, downstream of `is_bit_valid`).

## `validate_cast_and_convert_metadata`

https://github.com/google/zerocopy/blob/000318497d86599e24cecb66b945c86a53af3163/src/layout.rs#L444-L445

We should be able to make this work via a post-monomorphization error instead, and thus avoid a panic opportunity.

## `PointerMetadata::size_for_metadata`

https://github.com/google/zerocopy/blob/000318497d86599e24cecb66b945c86a53af3163/src/lib.rs#L719-L721

## `TryFromBytes::is_bit_valid`

https://github.com/google/zerocopy/blob/000318497d86599e24cecb66b945c86a53af3163/src/lib.rs#L1243-L1251

Now that const eval semantics are more nailed down, we can probably stop hedging that this might panic and just guarantee a post-monomorphization error.

Note that many panics are downstream of `is_bit_valid`. If we tackle this, we should make sure to remove panic documentation from all downstream functions.

## `round_down_to_next_multiple_of_alignment`

https://github.com/google/zerocopy/blob/000318497d86599e24cecb66b945c86a53af3163/src/util/mod.rs#L623-L624

We could benefit from a power-of-two witness type.

**Comment by jswrenn** [comment]
## `validate_cast_and_convert_metadata`

I don't think we can remove this one. `validate_cast_and_convert_metadata`'s sole callsite is in `Ptr::try_cast_into`. That method consumes a `meta: Option<U::PointerMetadata>` argument. This `meta` is then combined `U::LAYOUT` to create `layout: DstLayout`, on which  `validate_cast_and_convert_metadata` is called.

As a rule, we can only convert a panic to PME in scenarios where the PME condition is observable in a const context. This is a little too dynamic.

**Comment by joshlf** [comment]
Once our MSRV is high enough, maybe we can pass the `Layout` as a `KnownLayout` bound instead of as a value?

**Comment by clundin25** [comment]
Hello! I am hoping you can help me on this topic.

### Problem 

I'm working on migrating from zerocopy v0.6.6 to v0.8.6. This is an embedded RISCV codebase, that is strictly no-panic but I see the same behavior on my x86_64 Linux workstation.

I am blocked on `validate_cast_and_convert_metadata`. The compiler is not able to optimize this method out and thus retains a panic in the binary.

I'm hoping you might have some suggestions on how to tackle this.

### Examples

Here is a stripped down example of my code. This will introduce `validate_cast_and_convert_metadata` and a panic into the final binary.

```rust 
use core::mem::size_of;
use zerocopy::FromBytes;

fn main() {
    let q = [0x1, 0x1, 0x1, 0x1, 0x2, 0x2, 0x2, 0x2, 0xA];
    let count = q.len() / size_of::<u32>();
    let (word_buf, _suffix) = <[u32]>::ref_from_prefix_with_elems(&q, count).unwrap();

    let mut p = [0x1, 0x1, 0x1, 0x1, 0x2, 0x2, 0x2, 0x2, 0xA];
    let p_count = p.len() / size_of::<u32>();
    let (mut writeable_word_buf, _suffix) = <[u32]>::mut_from_prefix_with_elems(&mut p, p_count).unwrap();
}
```

This example will optimize `validate_cast_and_convert_metadata` out of the binary.

```rust 
use core::mem::size_of;
use zerocopy::FromBytes;

fn main() {
    let q = [0x1, 0x1, 0x1, 0x1, 0x2, 0x2, 0x2, 0x2, 0xA];
    let count = q.len() / size_of::<u32>();
    let (word_buf, _suffix) = <[u32]>::ref_from_prefix_with_elems(&q, count).unwrap();
}
```

Here are the profile options:

```
[profile.release]
panic = "abort"
lto = true
opt-level = "s"
codegen-units = 1
```

### Workarounds 

So far I have identified two workarounds with significant drawbacks.

#### Deserialize one u32 at a time for the immutable code path

I can deserialize one u32 at a time using `[read_from_bytes](https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html#method.read_from_bytes)`.

This has two major drawbacks:

* Performance takes a massive hit.
* I have to use a copy. Using the `ref` variant will pull in  `validate_cast_and_convert_metadata` and it's panic into the binary.

#### Only use the mutable API

I can refactor the code to only use the mutable `mut_from_prefix_with_elems` variant. This avoid introducing `validate_cast_and_convert_metadata` to the binary.

The major drawback here is that byte slices that were once immutable are now mutable across several layers of code, which could lead to future bugs. 

I don't think this workaround will get much buy in.

### Sources

I've posted an example [repo](https://github.com/clundin25/zero-copy-panic-example) on my GitHub profile. You can compile with `cargo build --release` and inspect the assembly to confirm the above behavior.

 Additionally I can share the production changes, since it's an open source project.

**Comment by clundin25** [comment]
This patch proposed by @jhand2 solves my problem. https://github.com/clundin25/zerocopy/commit/0b0d34570ab03eb3c07a55fe9c34ef69f809a5fc.

I'm going to spend some more time to verify it, but I wanted to share it to get your thoughts on this.

Would you consider adding this patch to 0.8?

**Comment by clundin25** [comment]
https://github.com/google/zerocopy/pull/1997 Resolves the issue I was facing.

**Comment by joshlf** [comment]
https://github.com/google/zerocopy/pull/1997 has been published in [0.8.8](https://crates.io/crates/zerocopy/0.8.8).
